File: app/src/main/java/com/bdd/workshop/ui/MainActivity.java
package com.bdd.workshop.ui;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bdd.workshop.App;
import com.bdd.workshop.R;
import com.bdd.workshop.model.Task;
import com.bdd.workshop.model.TaskManager;

import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity {

  private TaskManager taskManager;

  private RecyclerView recyclerView;

  @Override
  protected void onCreate(@Nullable final Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    taskManager = ((App) getApplication()).getTaskManager();

    setContentView(R.layout.activity_main);
    setSupportActionBar(findViewById(R.id.toolbar));

    recyclerView = findViewById(R.id.recyclerView);
    final TaskAdapter taskAdapter = new TaskAdapter(LayoutInflater.from(this));
    recyclerView.setAdapter(taskAdapter);
    recyclerView.setLayoutManager(new LinearLayoutManager(this));
    taskAdapter.setTasks(taskManager.getAllTasks());
  }


  static class TaskAdapter extends RecyclerView.Adapter<TaskViewHolder> {

    private final LayoutInflater inflater;
    private List<Task> items = Collections.emptyList();

    TaskAdapter(final LayoutInflater inflater) {
      this.inflater = inflater;
    }

    void setTasks(List<Task> tasks) {
      this.items = tasks != null ? tasks : Collections.emptyList();
      notifyDataSetChanged();
    }

    @Override
    public TaskViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
      return new TaskViewHolder(inflater.inflate(R.layout.item_task, parent, false));
    }

    @Override
    public void onBindViewHolder(final TaskViewHolder holder, final int position) {
      holder.updateTask(items.get(position));
    }

    @Override
    public int getItemCount() {
      return items.size();
    }
  }


  static class TaskViewHolder extends RecyclerView.ViewHolder {


    private TextView title;
    private TextView description;

    public TaskViewHolder(final View itemView) {
      super(itemView);
      title = itemView.findViewById(R.id.title);
      description = itemView.findViewById(R.id.description);
    }

    void updateTask(Task task) {
      title.setText(task.getTitle());
      description.setText(task.getDescription());
    }
  }
}

File: app/src/main/java/com/bdd/workshop/ui/MainActivity.kt
package com.bdd.workshop.ui

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bdd.workshop.App
import com.bdd.workshop.R
import com.bdd.workshop.model.Task
import com.bdd.workshop.model.TaskManager
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.item_task.view.*

class MainActivity : AppCompatActivity() {

  private lateinit var taskManager: TaskManager

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    taskManager = (application as App).taskManager

    setContentView(R.layout.activity_main)
    setSupportActionBar(toolbar)

    val taskAdapter = TaskAdapter(LayoutInflater.from(this))
    recyclerView.adapter = taskAdapter
    recyclerView.layoutManager = LinearLayoutManager(this)
    taskAdapter.tasks = taskManager.getAllTasks()

    fab.setOnClickListener {
      startActivity(Intent(this, AddTaskActivity::class.java))
    }
  }
}

private class TaskAdapter(val inflater: LayoutInflater): RecyclerView.Adapter<TaskViewHolder>() {

  var tasks: List<Task> = ArrayList()
    set(value) {
      field = value
      notifyDataSetChanged()
    }


  override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): TaskViewHolder =
      TaskViewHolder(inflater.inflate(R.layout.item_task, parent, false))

  override fun onBindViewHolder(holder: TaskViewHolder?, position: Int) {
    holder?.updateTask(tasks[position])
  }

  override fun getItemCount(): Int = tasks.size

}

class TaskViewHolder(view: View): RecyclerView.ViewHolder(view) {

  fun updateTask(task: Task) {
    itemView.title.text = task.title
    itemView.description.text = task.description
  }

}
